### ISSUE TITLE
Push image to registry error

### ISSUE THREAD

**Issue by ywei2017** [cloudfoundry-community/docker-registry-boshrelease#44]
We keep on getting errors when pushing an image to the registry. In particular with one called "pcf-tools". It is a big images with layers total more than 2 GB.

We are on the latest release 3.5.0, with an EMC based S3 backend. When look at the registry log, we see
`time="2019-08-06T19:46:44.792006652Z" level=error msg="response completed with error" err.code=unknown err.detail="s3aws: Path not found: /docker/registry/v2/repositories/pcf-tools/_uploads/96f47611-03ea-4b20-b125-5abed1a9b408/data" err.message="unknown error" go.version=go1.11 http.request.host=<redacted> http.request.id=99d86df5-058d-43ae-a02b-4af70be5c534 http.request.method=PATCH http.request.remoteaddr=<redacted> http.request.uri="/v2/pcf-tools/blobs/uploads/96f47611-03ea-4b20-b125-5abed1a9b408?_state=LWRdH3bu_aE-WtwdQrTcKAdeJqsmrf4E6PuiJi2y-MN7Ik5hbWUiOiJwY2YtdG9vbHMiLCJVVUlEIjoiOTZmNDc2MTEtMDNlYS00YjIwLWIxMjUtNWFiZWQxYTliNDA4IiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDE5LTA4LTA2VDE5OjQ2OjE4LjM3MDEyNDk3WiJ9" http.request.useragent="docker/18.03.1-ce go/go1.10.1 git-commit/v18.03.1-ce kernel/4.15.0-52-generic os/linux arch/amd64 UpstreamClient(Docker-Client/18.03.1-ce \(linux\))" http.response.contenttype="application/json; charset=utf-8" http.response.duration=81.236952ms http.response.status=500 http.response.written=193 vars.name=pcf-tools vars.uuid=96f47611-03ea-4b20-b125-5abed1a9b408`

We were on 3.2.x, and though maybe the new release would solve the issue. Unfortunately it didn't.

As background, we used to have no issue pushing the image. Then later it began to fail randomly with increased frequency. Now it consistently fails :-(

Would appreciate any ideas to further troubleshoot, or solve the issue.

Thanks
Yansheng

**Comment by jhunt** [comment]
Is it possible that the EMC S3 work-alike is having issues?

Can you share a (possibly redacted) version of your deployment manifest?

**Comment by ywei2017** [comment]
Hi, @jhunt , no problem, see the full redacted manifest. A few conflicting notes.
- We have no problem to push images in stage environment.  Only prod has issues. 
- Maybe the prod and stage EMC S3 cluster are configured slightly differently. But I don't know how/what to check. Any suggestion to ascertain the S3 cluster will be appreciated.
- The prod S3 cluster is used for our PCF backend blobstore, and we never had a single issue.

`
instance_groups:
- azs:
  - z1
  env:
    bosh:
      password: <redacted>
  instances: 2
  jobs:
  - name: registry
    release: docker-registry
  - name: proxy
    release: docker-registry
  - name: cache
    release: docker-registry
  name: docker-registry
  networks:
  - name: repo
    static_ips:
    - <ip1-redacted>
    - <ip2-redacted>
  properties:
    docker:
      cache:
        host: 127.0.0.1
        redis:
          maxmemory: 2g
      loglevel: info
      proxy:
        auth_basic:
          concourse: '{PLAIN}0fTOz7R_hjJFaBdQ9vEhQSjZ6Qq0VQ5s'
          docker: '{PLAIN}0fTOz7R_hjJFaBdQ9vEhQSjZ6Qq0VQ5s'
        only_auth_for_admin: false
        ssl:
          cert: |-
            -----BEGIN CERTIFICATE-----
            <redacted>
            -----END CERTIFICATE-----
          key: |-
            -----BEGIN PRIVATE KEY-----
            <redacted>
            -----END PRIVATE KEY-----
      registry:
        bind: 127.0.0.1
        endpoint: https://<external_host_name>:443  #F5 load balancer infront of ip1 and ip2
        storage:
          s3:
            accesskey: <redacted>
            bucket: <redacted>
            chunksize: 5242880
            multipartcopychunksize: 33554432
            multipartcopymaxconcurrency: 100
            multipartthresholdcopysize: 33554432
            region: default
            regionendpoint: <redacted - EMC S3 bucket URL>
            rootdirectory: null
            secretkey: <redacted>
            v4auth: true
        storage_disable_redirect: true
  stemcell: ubuntu
  vm_extensions: []
  vm_type: large.disk
name: docker-registry
releases:
- name: docker-registry
  version: latest
stemcells:
- alias: ubuntu
  os: ubuntu-xenial
  version: latest
update:
  canaries: 1
  canary_watch_time: 1000-60000
  max_in_flight: 50
  serial: false
  update_watch_time: 1000-60000

`

**Comment by jhunt** [comment]
Is it possible that you are running into disk space issues in the fronting nginx proxy (the `proxy` job on the `docker-registry` instance group)?  Is there any rhyme or reason to the failures, specifically based on the size of the image layers being pushed?

**Comment by ywei2017** [comment]
I checked the disk usage over the past 2 weeks. Ephemeral disk has been about 1% used, and persistent disk at 50%. So it's not disk space issue.

While the image is relatively large (about 2GB total), the individual layers aren't that big.

**Comment by jhunt** [comment]
I've seen similar issues before where nginx is trying to copy the request body into a temporary directory, but lacks enough free space to completely spool the request payload.  When this happens, you get HTTP 500 errors with little context, but because nginx cleans up after itself in these situations, disk usage looks great both before and after.

(Also, given the speed of modern disks and networks, it's quite difficult to catch the out-of-space error as a human)

Two things I would try (in increasing order of complexity):

1. Remove (or bypass) the proxy and go straight to the registry application with a push.  This eliminated nginx entirely and will help to pinpoint the problem.  If it works without the proxy, nginx is probably the root cause.  If the issue persists, we'll need to look elsewhere.

2. `strace` each nginx worker PID on the BOSH VM.  On particularly busy registries this can be quite a lot of output, but if nginx is indeed the problem, this will find it every single time.  You'll see an `open(2)` call made against a path called something like "client_body_tmp", followed by a series of `write(2)` calls that culminate in an errno = `ENOSPC` situation.  If this is what's happening, you'll want to increase the size of the disk that the client_body_tmp file (whatever it is called) is being created.  Make sure that you strace _all_ of the `nginx worker` processes on the box, since nginx passes incoming connections to the whole pool and you can't easily predict _which_ worker will get the request you care about.

I just did this with a BOSH Director Blobstore (also nginx) that was exhibiting the exact same problems -- semi-intermittent failures, plenty of free store and ephemeral disk, etc.

**Comment by ywei2017** [comment]
Hi, @jhunt ,

The following is the config for nginx (default), by looking at it, there is no directive for caching, does this mean caching is turned off [by default](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache)?

`docker-registry/c36e457e-a795-46e3-98eb-e4fb5486e138:/var/vcap/jobs/proxy/bin# cat /var/vcap/jobs/proxy/config/nginx.conf
user vcap;

pid /var/vcap/sys/run/proxy/nginx.pid;
error_log /var/vcap/sys/log/proxy/error.log error;
working_directory /var/vcap/data/tmp;
worker_processes auto;
worker_priority 0;

events {
    worker_connections 8192;
    use epoll;
}

http {
    # Basic settings
    server_tokens off;
    sendfile off;
    tcp_nopush on;
    tcp_nodelay on;
    keepalive_timeout 75s 20s;
    keepalive_requests 512;
    connection_pool_size 256;
    client_max_body_size 0;
    chunked_transfer_encoding on;

    # Log format
    log_format timed_combined '$remote_addr - $remote_user [$time_local] "$request" $status $body_bytes_sent "$http_referer" "$http_user
_agent" $request_time $upstream_response_time $pipe';

    access_log /var/vcap/sys/log/proxy/proxy-access.log timed_combined;
    error_log  /var/vcap/sys/log/proxy/proxy-error.log error;
        # Recommendations from https://raymii.org/s/tutorials/Strong_SSL_Security_On_nginx.html
        ssl_protocols TLSv1.1 TLSv1.2;
        ssl_ciphers 'EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH';
        ssl_prefer_server_ciphers on;
        ssl_session_cache shared:SSL:10m;

        ssl_certificate     /var/vcap/jobs/proxy/tls/cert.pem;
        ssl_certificate_key /var/vcap/jobs/proxy/tls/key.pem;


    ## Set a variable to help us decide if we need to add the
    ## 'Docker-Distribution-Api-Version' header.
    ## The registry always sets this header.
    ## In the case of nginx performing auth, the header will be unset
    ## since nginx is auth-ing before proxying.
    map $upstream_http_docker_distribution_api_version $docker_distribution_api_version {
        'registry/2.0' '';
        default registry/2.0;
    }

    upstream docker {

        server 127.0.0.1:5000;

    }
    ## If $docker_distribution_api_version is empty, the header will not be added.
    ## See the map directive above where this variable is defined.
    add_header 'Docker-Distribution-Api-Version' $docker_distribution_api_version always;
    proxy_set_header Host              $host;
    proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
    proxy_set_header X-Real-IP         $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Original-URI    $request_uri;
    proxy_pass_header Server;
    proxy_read_timeout 3600;


    server {
        server_name proxy_admin;

            listen *:443 default_server;
            ssl on;

        location /v2/ {
            # Do not allow connections from docker 1.5 and earlier
            # docker pre-1.6.0 did not properly set the user agent on ping, catch "Go *" user agents
            if ($http_user_agent ~ "^(docker\/1\.(3|4|5(?!\.[0-9]-dev))|Go ).*\$" ) {
                return 404;
            }

                # To add basic authentication to v2 use auth_basic setting.
                auth_basic "Registry Realm";
                auth_basic_user_file auth/registry;

            proxy_pass http://docker;
        }
    }



}
`

**Comment by ywei2017** [comment]
Hi, @jhunt , took me a few days to follow through on the procedure. Due to firewall restrictions, etc, I finally was able to do it by installing the docker release on the same registry host, so I can do "docker push", targetting the registry at localhost, and bypassing the proxy.

I this scenario, I got the same error. and I am pasting the complete message from a particular failed layer:
`time="2019-08-13T19:29:45.733180524Z" level=debug msg="s3aws.Writer("/docker/registry/v2/repositories/pcf-tools/_uploads/09fa98d9-fc08-4612-8052-94c5e0df248a/data", false
)" go.version=go1.11 http.request.host="127.0.0.1:5000" http.request.id=9881fa07-012d-4cf5-8847-8136fedbbbc7 http.request.method=POST http.request.remoteaddr="127.0.0.1:5
2074" http.request.uri="/v2/pcf-tools/blobs/uploads/" http.request.useragent="docker/1.13.1 go/go1.7.5 git-commit/092cba3 kernel/4.15.0-55-generic os/linux arch/amd64 Ups
treamClient(Docker-Client/1.13.1 \(linux\))" trace.duration=73.842548ms trace.file="/var/vcap/packages/distribution/src/github.com/docker/distribution/registry/storage/dr
iver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Writer" trace.id=9dd3ae16-9e36-40fb-9871-b9f4cf4fd06d trace.line=142 va
rs.name=pcf-tools
time="2019-08-13T19:29:46.159683693Z" level=debug msg="s3aws.Writer("/docker/registry/v2/repositories/pcf-tools/_uploads/09fa98d9-fc08-4612-8052-94c5e0df248a/data", true)
" go.version=go1.11 http.request.host="127.0.0.1:5000" http.request.id=e2c0b3eb-37d1-4179-87cf-cabe464fd894 http.request.method=PATCH http.request.remoteaddr="127.0.0.1:5
2084" http.request.uri="/v2/pcf-tools/blobs/uploads/09fa98d9-fc08-4612-8052-94c5e0df248a?_state=oxVpbzLAAm5CBtSg4oJGp71sPufPg3NmLF8_M_qOm8N7Ik5hbWUiOiJwY2YtdG9vbHMiLCJVVU
lEIjoiMDlmYTk4ZDktZmMwOC00NjEyLTgwNTItOTRjNWUwZGYyNDhhIiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDE5LTA4LTEzVDE5OjI5OjQ1LjYxNDU4NjM5M1oifQ%3D%3D" http.request.useragent="docker
/1.13.1 go/go1.7.5 git-commit/092cba3 kernel/4.15.0-55-generic os/linux arch/amd64 UpstreamClient(Docker-Client/1.13.1 \(linux\))" trace.duration=29.642846ms trace.file="
/var/vcap/packages/distribution/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/drive
r/base.(*Base).Writer" trace.id=df316a9f-5745-4a9c-bd4c-9d0ec1d54277 trace.line=142 vars.name=pcf-tools vars.uuid=09fa98d9-fc08-4612-8052-94c5e0df248a
time="2019-08-13T19:29:46.159725725Z" level=error msg="error resolving upload: s3aws: Path not found: /docker/registry/v2/repositories/pcf-tools/_uploads/09fa98d9-fc08-46
12-8052-94c5e0df248a/data" go.version=go1.11 http.request.host="127.0.0.1:5000" http.request.id=e2c0b3eb-37d1-4179-87cf-cabe464fd894 http.request.method=PATCH http.reques
t.remoteaddr="127.0.0.1:52084" http.request.uri="/v2/pcf-tools/blobs/uploads/09fa98d9-fc08-4612-8052-94c5e0df248a?_state=oxVpbzLAAm5CBtSg4oJGp71sPufPg3NmLF8_M_qOm8N7Ik5hb
WUiOiJwY2YtdG9vbHMiLCJVVUlEIjoiMDlmYTk4ZDktZmMwOC00NjEyLTgwNTItOTRjNWUwZGYyNDhhIiwiT2Zmc2V0IjowLCJTdGFydGVkQXQiOiIyMDE5LTA4LTEzVDE5OjI5OjQ1LjYxNDU4NjM5M1oifQ%3D%3D" http.
request.useragent="docker/1.13.1 go/go1.7.5 git-commit/092cba3 kernel/4.15.0-55-generic os/linux arch/amd64 UpstreamClient(Docker-Client/1.13.1 \(linux\))" vars.name=pcf-
tools vars.uuid=09fa98d9-fc08-4612-8052-94c5e0df248a
time="2019-08-13T19:29:46.159793931Z" level=error msg="response completed with error" err.code=unknown err.detail="s3aws: Path not found: /docker/registry/v2/repositories
/pcf-tools/_uploads/09fa98d9-fc08-4612-8052-94c5e0df248a/data" err.message="unknown error" go.version=go1.11 http.request.host="127.0.0.1:5000" http.request.id=e2c0b3eb-3
7d1-4179-87cf-cabe464fd894 http.request.method=PATCH http.request.remoteaddr="127.0.0.1:52084" http.request.uri="/v2/pcf-tools/blobs/uploads/09fa98d9-fc08-4612-8052-94c5e
0df248a?_state=oxVpbzLAAm5CBtSg4oJGp71sPufPg3NmLF8_M_qOm8N7Ik5hbWUiOiJwY2YtdG9vbHMiLCJVVUlEIjoiMDlmYTk4ZDktZmMwOC00NjEyLTgwNTItOTRjNWUwZGYyNDhhIiwiT2Zmc2V0IjowLCJTdGFydGV
kQXQiOiIyMDE5LTA4LTEzVDE5OjI5OjQ1LjYxNDU4NjM5M1oifQ%3D%3D" http.request.useragent="docker/1.13.1 go/go1.7.5 git-commit/092cba3 kernel/4.15.0-55-generic os/linux arch/amd6
4 UpstreamClient(Docker-Client/1.13.1 \(linux\))" http.response.contenttype="application/json; charset=utf-8" http.response.duration=61.079144ms http.response.status=500
http.response.written=193 vars.name=pcf-tools vars.uuid=09fa98d9-fc08-4612-8052-94c5e0df248a
`

**Comment by ywei2017** [comment]
Hi, @jhunt , more on the troubleshooting.
* I changed the storage to local filesystem, and "docker push" was successful the firs time. So that concludes that it's somewhat specific to S3.
* Then I changed the chunksize from the default "5242880" to "524288000" (100x bigger)

That seems to do the trick. I will know better once we do it a few more times. So time will definitely tell. For now, I think we can close the case, since I don't think there is anything inherently wrong with the bosh release. 

Thanks for the kind help. Very much appreciate it.
Yansheng
